cmd/ltkeycloak/main.go

```diff
@@ -0,0 +1,468 @@
+				if err := keycloakClient.AddUserToGroup(ctx, token.AccessToken, deploymentConfig.ExternalAuthProviderSettings.KeycloakRealmName, *user.ID, keycloakMigratedGroup); err != nil {
+					return fmt.Errorf("failed to mark user migrated in keycloak: %w", err)
+				}
+			}
+
+			slog.Info("migrated user", slog.String("username", *user.Username))
+
+			if len(users) == 0 {
+				break
+			}
+
+			start += perPage
+		}
+
+		return nil
+	}
+}
+
+func RunSyncFromMattermostCommandF(cmd *cobra.Command, _ []string) error {
+	started := time.Now()
+
+	ltConfigPath, err := cmd.Flags().GetString("config")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	// Guess where the deployment configuration is located
+	deploymentConfig, err := deployment.ReadConfig(filepath.Join(filepath.Dir(ltConfigPath), "deployer"+filepath.Ext(ltConfigPath)))
+	if err != nil {
+		return fmt.Errorf("failed to read deployment configuration: %w", err)
+	}
+
+	dryRun, err := cmd.Flags().GetBool("dry-run")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	mmClient := model.NewAPIv4Client("http://" + deploymentConfig.SiteURL)
+
+	_, _, err = mmClient.Login(cmd.Context(), deploymentConfig.AdminEmail, deploymentConfig.AdminPassword)
+	if err != nil {
+		return fmt.Errorf("failed to login to mattermost: %w", err)
+	}
+
+	var keycloakHost string
+	keycloakHost, err = cmd.Flags().GetString("keycloak-host")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	// Use the terraform output terraform host if a manual one is not provided. Useful for development.
+	if keycloakHost == "" {
+		t, err := terraform.New("", *deploymentConfig)
+		if err != nil {
+			return fmt.Errorf("failed to create terraform client: %w", err)
+		}
+		terraformOutput, err := t.Output()
+		if err != nil {
+			return fmt.Errorf("failed to get terraform output: %w", err)
+		}
+		if terraformOutput.KeycloakServer.PublicDNS == "" {
+			return fmt.Errorf("keycloak database cluster not found in terraform output")
+		}
+		keycloakHost = terraformOutput.KeycloakServer.PublicDNS
+	}
+
+	keycloakRealm, err := cmd.Flags().GetString("keycloak-realm")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	userPassword, err := cmd.Flags().GetString("set-user-password-to")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	keycloakClient := gocloak.NewClient(keycloakHost)
+	ctx := context.Background()
+
+	keycloakToken, err := keycloakClient.LoginAdmin(
+		ctx,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminUser,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminPassword,
+		"master", // TODO: Allow specifying the master realm
+	)
+	if err != nil {
+		return fmt.Errorf("failed to refresh keycloak token: %w", err)
+	}
+
+	usersChan := make(chan *model.User, 10000)
+	usersTxtChan := make(chan string, 10000)
+	doneChan := make(chan struct{})
+	workers := 4
```

Comment:
Bad practice to hardcode workers. Should be always `runtime.NumCPU()` capped to a limit.